### ISSUE TITLE
Show important peaks sooner and fewer minor peaks on z11

### ISSUE THREAD

**Issue by jeisenbe** [gravitystorm/openstreetmap-carto#3722]
### Expected behavior
Prominent and isolated peaks should be shown at mid zoom levels. It's common for large-scale maps to show the tallest mountain peak in a country, thus the most important mountain peaks should be shown as early as z8, when rivers are first shown. 

Minor sub-peaks on the sides of larger mountains and insignifcant hills should not be shown until high-mid zoom levels (eg z14)

The names and elevations of important mountains should be rendered with priority. They should not be blocked by nearby minor peaks

### Actual behavior
All natural=peak features are shown from z11 without names or elevation, and the names are shown at z10 without clear prioritization of prominent peaks. In areas where many small peaks have been mapped, the map is full of peak symbols at z11

While the icons of taller peaks are prioritized, the name label of a tall mountain can be blocked by a minor subsidiary peak; eg the tiny "South Summit" of Mt Everest blocks display of the main peak's name label at several zoom levels. 

### Possible Solutions

1) Use existing OSM tags, including name and elevation tags, to simply filter out less important peaks (i.e. those lacking names, elevation, or both). 
- This can reduce the number of peaks in some areas, but many areas would still have too many at z11, and it would not be easy to filter for the 1% of most important peaks which could be shown at lower zoom levels.

2) Calculate an importance value for each peak based on the distance to the next higher peak
- This is similar to isolation, but would only look at the distance to other peaks (and saddles?) tagged with elevation
- Could be sufficient to show some very important peaks sooner, but would not work well in areas where elevation is not tagged for many peaks

3) Preprocess peaks from OSM database with a DEM (digital elevation model) to determine the isolation of each peak, and use this to help rank importance (along with name and elevation)
- Requires preprocessed data stored elsewhere or preprocessing along with a DEM at time of rendering
- Should allow a fairly robust importance rating for all peaks in the OSM database, including those that lack elevation tags
- Could be done frequently; daily basis

4) Preprocess peaks with a DEM to calculate prominence
- More complicated that finding isolation, because this requires creating a model of all ridge lines and saddles in addition to the location of peaks. However, prominence is considered a better importance rating for peaks by most mountaineers and climbers. 
- Probably would need to be done infrequently; once per month, with results stored in a separate file (eg at osmdata.openstreetmap.de)

### Links and screenshots illustrating the problem

![](https://user-images.githubusercontent.com/1190866/30525409-e84d8c08-9bba-11e7-
8313-e3f43705c575.png)
- The name and elevation of Mt Everest are blocked by the icon for South Peak (a tiny bump on the ridge with prominence of only 10 meters)

![many-peaks-cimahi-11:-6 9384:107 4223](https://user-images.githubusercontent.com/42757252/54795771-d5f33c00-4c90-11e9-98c4-12e2fb4ee3cf.png)
https://www.openstreetmap.org/#map=11/-6.9384/107.4223
- z11 East of Bandung in Java, many peaks are shown, but it isn't clear what is the tallest mountain, or where the main ridgeline is located.

![many-peaks-11:48 9651:19 2563](https://user-images.githubusercontent.com/42757252/54795764-d2f84b80-4c90-11e9-9347-871199c6a110.png)
https://www.openstreetmap.org/#map=11/48.9651/19.2563
- z11 Many peaks are shown at z11 near Ruzomberok in Slovakia, but most do not add clear information to the map

![sulawesi-peaks-11:-3 2416:119 0774](https://user-images.githubusercontent.com/42757252/54795770-d5f33c00-4c90-11e9-8996-44f1e69f0d7e.png)
https://www.openstreetmap.org/#map=11/-3.2416/119.0774
- z11 in Sulawesi, Indonesia. Showing all of these peaks equally at z11 is not very helpful.

Zooming out to z10 shows no peaks. Mauna Kea in Hawaii is the tallest peak for 1000s of kilometers in all directions, and an elevation of 4205 meters, but is not shown at z10.
https://www.openstreetmap.org/node/2189450439#map=12/19.82069/-155.46804
![n-hawaii-z10-current](https://user-images.githubusercontent.com/42757252/54796086-228b4700-4c92-11e9-99b2-78c6e473826d.png)

**Comment by jeisenbe** [comment]
The idea of using external data (e.g. a DEM for elevation data) was discussed in #3720 where it was decided that using data from other crowdsourced databases (e.g. Wikidata) to find pre-calculated prominence or isolation values would not be a good idea.

**Comment by matkoniecz** [comment]
> Use existing OSM tags, including name and elevation tags, to simply filter out less important peaks (i.e. those lacking names, elevation, or both).

https://taginfo.openstreetmap.org/tags/natural=peak#combinations - I am quite surprised that 28% of peaks has no specified `name` tag. It is quite weird and worth investigating.

**Comment by matkoniecz** [comment]
> The name and elevation of Mt Everest are blocked by the icon for South Peak (a tiny bump on the ridge with prominence of only 10 meters)

Note that peaks are already rendered with priority, by elevation.

Adding new way of calculating priority will not help at all in this case.

Fixing this requires making possible to render high priority label before peak icon with a lower priority,

**Comment by jeisenbe** [comment]
> Peaks are already rendered with priority, by elevation.
> Adding new way of calculating priority will not help at all in this case.
> Fixing this requires making possible to render high priority label before peak icon with a lower priority

It would solve the problem for low-mid zoom levels; for example, only peaks with isolation >10 km or prominence >200m might be shown at z9 to z12, thus Everest would be shown and South Peak would not. 

It's true that additional changes would be required to fix the problem at mid-high zoom levels in some case, but having a better prioritization system could still be helpful. Checking for isolation in particular can make certain that there is enough room for a normal sized name label for the peak with higher priority.

**Comment by imagico** [comment]
> Fixing this requires making possible to render high priority label before peak icon with a lower priority,

Note this is a problem that is not specific to peaks, you have it also with higher zoom level POIs (mostly the other way round - with labels disappearing as you zoom in because new icons turn up).  This is one of the major problems i mentioned that need fixing in #3635.

**Comment by jeisenbe** [comment]
@imagico - would this be proper data to include at osmdata.openstreetmap.de (a file including all OSM peaks with their isolation or prominence, calculated automatically on a frequent basis)? I suppose there would need to be scripts and an example of the results first.

**Comment by imagico** [comment]
As said in another context i would suggest to discuss this based on a specific process once it exists and not in abstract form for something yet to be developed.

As also said i would in general be very careful with adding dependencies on external data to this style - no matter how it is produced.  There would need to be a clear justification in terms of the goals we have that goes beyond a simple _it looks better_.

**Comment by Tomasz-W** [comment]
@jeisenbe Can you do something about this issue? Ir causes real mess in some places:

https://www.openstreetmap.org//#map=11/50.2006/16.6834
https://www.openstreetmap.org//#map=11/47.3914/7.6877
https://www.openstreetmap.org//#map=11/45.7761/14.1133

**Comment by matkoniecz** [comment]
@Tomasz-W Please, avoid +1 comments. Reactions added by Github are much better way to express that  ( see https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/ guide for how one may use them )

**Comment by Tomasz-W** [comment]
@matkoniecz Please avoid looking for a problem everywhere. Commenting abandoned tickets is a common way to suggest maintainers which issues are more important and which are less. Reactions don't provide notifications while comments do.

**Comment by matkoniecz** [comment]
> Reactions don't provide notifications while comments do.

That is the main reason why +1 comments are awful and unwanted.

**Comment by jeisenbe** [comment]
Here's another example, from all the way up at 53 degrees north in Wales; there would only be 60% as much space between peaks at the equator for an area mapped similarly, leading to even more collisions:

![z11-bala-too-many-peaks](https://user-images.githubusercontent.com/42757252/64261877-325cc600-cf68-11e9-8cb1-575d40f486bf.png)

Is it possible to calculate the distance to the nearest `natural=peak` node in SQL? 

That would be a simpler version of "isolation" which could work well to avoid rendering minor peaks at z11 and z12.

**Comment by kennykb** [comment]
"Feature density" is something that would be nice to have for all POI's, but I haven't found anything better yet than what we already do: try to order features by some sort of importance, and then suppress ones that would paint over existing ones - that's done in Mapnik of course.

"Find the nearest neighbouring peak" is doable in PostGIS but it's pretty horrific. It's a special case of "find the K nearest neighbours", and it really needs you to create an index to support the query, since the search HAS to be with the `<->` operator and there _has_ to be a `LIMIT` clause in the subquery. It's also a very bad thing to do in bulk; if you have a set (like the peaks in a super-tile) and need nearest-neighbor for all of them, the time to find the neighbours goes up faster than linearly. It's almost certainly better to do some sort of processing on top of `ST_DelaunayTriangles` instead. In any case, this will _not_ be a fast query. It's rather a difficult question that we're asking, and the answer will be corrrespondingly expensive to compute.

I've experimented with some other approaches in maps that I render. One thing that, while not entirely satisfactory, has at least improved matters for me is to start rendering 'buillt' PoI's at a different zoom level depending on whether the area is 'urban' or 'rural' That mans that I can continue to show availability of services at low zoom in the countryside without turning the cities into totally unreadable clutter. Here, the working definition of 'urban' is 'intersects the polygon for an urban area in Natural Earth's 1:10 000 000-scale data. The query to return PoI's looks like
```
       (SELECT way, aerialway, aeroway, amenity, highway, leisure,
               man_made, "natural", "power", railway, religion,
               tourism, waterway, (u.ogc_fid IS NOT NULL) AS urban
	FROM
        (SELECT ST_PointOnSurface(way) AS way,
                aerialway, aeroway, amenity, highway, leisure,
                man_made, "natural", power, railway, religion, 
                tourism, waterway
         FROM &db_osm_polygon_table;
         WHERE
	   (aerialway IN ('station')
	    OR aeroway IN ('aerodrome', 'airport')
            -- about a dozen more OR conditions go here!
          )
	  AND ST_Intersects(way, ST_SetSRID(!bbox!,3857))

         UNION
         SELECT way,
	        aerialway, aeroway, amenity, highway, leisure, 
                man_made, "natural", power, railway, religion, tourism, waterway
         FROM &db_osm_point_table;
         WHERE
	   (aerialway IN ('station')
	    OR aeroway IN ('aerodrome', 'airport')
            -- about a dozen more OR conditions go here!
          )
	  AND ST_Intersects(way, ST_SetSRID(!bbox!,3857))) t
	  LEFT JOIN ne_10m_urban_areas u
	  ON ST_INTERSECTS(t.way, u.wkb_geometry)
	) AS Result
```
For maintainability, I really need to rewrite that monster with `WITH` clauses instead of nested `SELECT`'s.

I don't find the fact that the same type of PoI might appear at a different zoom level depending on context to be terribly distracting. Most users of the maps simply zoom until enough detail appears, and decluttering only helps.  

The 'urban-rural' distinction is, to me, even more critical than the latitude-dependent one. There would be a difference of almost exactly 1 zoom level in going from Quito to Oslo.  Two zoom levels would go to the latitude of Longyearbyen, and there's very little mapping in the extreme polar regions. (We'd need a whole different approach for mapping up there - polar-stereographic perhaps.) By contrast, I find that to get a good feature density in the countryside as opposed to the cities, a 2-3 zoom level difference is routine.

The use of Mercator at extremely low (z3 for instance) is certainly of tremendous political significance, because it does overemphasize the northern latitudes and produce a Eurocentric map. It's of little consequence at high zoom levels, where the scale distortion is no longer apparent and you're left with a much less consequential amount of feature clutter. 

Since there's really no good answer if you presuppose a rectangular canvas that will be recursively divided into tiles of progressively smaller size, the map scale question may not have a general answer at our current level of technology.

**Comment by jeisenbe** [comment]
> urban-rural distinction

Peaks are too dense even in very rural areas and at high latitudes, like the example at 53 north in Wales, so neither a change in map projection or differentiating rural and urban areas would help with this particular issue. 

> really no good answer if you presuppose a rectangular canvas that will be recursively divided into tiles of progressively smaller size

OT: While it would be difficult to change this map style away from mercator (since all existing links to maps on the openstreetmap.org main page assume this projection) there's no technical reason that you can't render your own web maps in a rectangular equal-area projection: https://en.wikipedia.org/wiki/Cylindrical_equal-area_projection

**Comment by c933103** [comment]
Could improvement be made by introducing new tags, like `prominence=\*` (although I am not sure how can such data be freely obtained/verified), or `main_peak=yes`, although I am not sure about how can you define it.

**Comment by StyXman** [comment]
Just some random thoughts here: Take in account that we use Mapnik, and we're limited by it.

In particular, peak symbols and their names are handled in two different layers; first the symbols, then the names. Also symbols and text fight for the same space, and because symbols are drawn first, they're mostly not being drawn in places with dense peaks. That's actually what's going on with your sshot about the Everest: its label has no space because the symbol for the South Peak is already there.

I tried by putting the peaks on their own layer and try to add some margin, bur `MarkerSymbolizer` does not have a `margin` attribute. I tried to let text `overlap`, but it overlaps itself and becomes unreadable. I wish Mapnik was more maintained and/or easier to modify.

**Comment by dch0ph** [comment]
As we are discussing labelling of natural features, I would like to give this issue a bump.

It is clearly a problem to show all `natural=peak` at Z11 (lots of examples above). 

Filtering by prominence will be hard, and frankly not very useful for a non-topographic map (without contours / hillshading). What's the point of showing a small triangle at the top of Mt. Everest with no indication (not even elevation) that it is a major geographical feature?

How about rendering `natural=peak` from Z13? This would be much more consistent with other "point" geographical features e.g. `natural=saddle` (only rendered from Z15).

For example, not this:

[Z11]
![Image](https://github.com/user-attachments/assets/d59f8365-28af-4ec6-8180-9a56a80bc6cc)

But start with this:

[Z13]
![Image](https://github.com/user-attachments/assets/e9b86742-d027-40eb-82e2-2d16b807881f)

**Comment by imagico** [comment]
> Filtering by prominence will be hard, 

Actually filtering is not hard at all, determining prominence is.  But then again - topographic prominence is not a very good measure of cartographic importance anyway.  A peak of a few hundred meters elevation might have a negligible prominence but is still by far the highest peak within hundreds of kilometers - and as such cartographically significant.

> and frankly not very useful for a non-topographic map (without contours / hillshading). What's the point of showing a small triangle at the top of Mt. Everest with no indication (not even elevation) that it is a major geographical feature?

If only the most significant peaks are shown (by some meaningful measure of significance) the indication will be the fact that it is rendered.  Showing the highest peaks of a mountain range with symbols at small scales is helpful to understand the topography - especially in a map that otherwise does not depict relief in any way.

As @StyXman pointed out the main difficulty is the constraints in symbol and label prioritization - partly due to the limitations of Mapnik, partly due to our messed up system of POI rendering (#3880).  Calculating a meaningful importance rating for peaks in preprocessing (either with a separate process or through osm2pgsql) is fairly simple compared to that.

This issue remaining open means we continue to welcome work in that direction.  This could be a starting point for developing a preprocessing strategy in general - see #4901.

**Comment by c933103** [comment]
I just read that some organization define a "mountain" to be a feature with prominence of more than 300 meters, and that there are about 191,000 of them around entire earth.
I feel like this could be a good compromises for displaying the symbol of peak at zoom level 9, and together with the name of the peaks at zoom level 10.
It wouldn't change existing rendering at zoom level 11 or below, therefore peaks without prominence data, or lower hills that are perceived to be significant on relatively flat land, can still be rendered at zoom level 11 just like now.

**Comment by dch0ph** [comment]
I didn't mean to argue the principle of showing important peaks early. I agree with "_[s]howing the highest peaks of a mountain range with symbols at small scales is helpful to understand the topography_", but highlighting important peaks will necessarily involving labelling them, otherwise they are anonymous orange/brown triangles.

But showing labelling important peaks would also mean not showing unlabelled unimportant peaks at low zooms, until the point where all peaks can be labelled (Z13+).

The unlabelled peaks at Z11/Z12 are not providing useful cartographic information, and it would seem a quick win to suppress them ahead of a more comprehensive solution to this issue.

**Comment by imagico** [comment]
> The unlabelled peaks at Z11/Z12 are not providing useful cartographic information

That is what i put into question in my comments.  Showing the most important peaks (which some meaningful measure of cartographic importance) adds meaningful information on the topography, in particular also without labels.

The problem we have at z11/12 is not that peaks are shown, it is that they are shown indiscriminately.

The question if the starting zoom levels for peak symbols or labels should be changed until a solution to this issue is available is another story - but that is not going to solve this issue - and, considering that any such hard cutoff will inevitably either discriminate against low or high latitudes i would rather see this issue solved.

**Comment by dch0ph** [comment]
May be worth noting that [OpenTopoMap uses "isolation"](https://github.com/der-stefan/OpenTopoMap/blob/master/mapnik/tools/update_isolations.sh) on the lines of alternatives 2/3 at the top of the issue.

This involves an external script being run periodically to update new isolation values, which may not be acceptable.

I don't have a good sense of the additional options provided by the flex input of osm2pgsql, but it looks that the "GIS capabilities" at import are deliberately restricted.

A generated column may be an alternative to an external script [import database, calculate new column of metric X, which can be approximately updated if new peaks are added.] It's just unfortunate that the support for generated columns in osm2pgsql is so qualified ("Don't rely on this behaviour.").
